lib/DB/Schema/Result/UserSet/HWSet.pm

```diff
@@ -1,31 +1,23 @@
 package DB::Schema::Result::UserSet::HWSet;
-use base qw/DB::Schema::Result::UserSet/;
+
 use strict;
 use warnings;
-
 use feature 'signatures';
-no warnings qw(experimental::signatures);
+no warnings qw/experimental::signatures/;
 
+use base qw/DB::Schema::Result::UserSet/;
 use DB::Schema::Result::ProblemSet::HWSet;
 
-sub valid_dates ($=) {
-	return DB::Schema::Result::ProblemSet::HWSet::valid_dates();
-}
-
-sub required_dates ($=) {
-	return DB::Schema::Result::ProblemSet::HWSet::required_dates();
-}
-
-sub optional_fields_in_dates ($=) {
-	return DB::Schema::Result::ProblemSet::HWSet::optional_fields_in_dates();
+sub valid_fields ($self, %args) {
```

Comment:
I don't think the valid_fields, required and additional_validation are needed in the subclasses because of the new way you create the subclasses. Testing with print statements inside these are never printed and everything passes if removed.

Reply:
I do believe you're right -- the UserSet subclasses can essentially disappear at this point. We may even be able to remove the `inflate_result` method as well... (that was only there because of the use of `->new`)

Reply:
All of tests run fine without that.  I thought that was key to handling the subclasses of UserSets.  I have to think on this.

Reply:
UserSet params are validated by merging them onto their *existing* related ProblemSet (subclassed), which knows how to validate itself. The return on that validation is the 'cleaned' params (only those that differ from the related ProblemSet). The UserSet is then created from that 'clean' hash of params.

Nowhere in the process do we need to subclass a UserSet. So, it's safe, I presume, to dump all UserSet subclasses and associated methods.

Reply:
Essentially, whenever a UserSet is requested, we should be returning a ProblemSet 'decorated' with the UserSet. The same should be applied when requesting a UserProblem (return a 'decorated' SetProblem).